src/core/tsi/alts/handshaker/alts_handshaker_client.cc

```diff
@@ -358,6 +335,164 @@ static tsi_result make_grpc_call(alts_handshaker_client* c, bool is_start) {
   return TSI_OK;
 }
 
+// TODO(apolcyn): remove this global queue when we can safely rely
+// on a MAX_CONCURRENT_STREAMS setting in the ALTS handshake server to
+// limit the number of concurrent handshakes.
+namespace {
+
+class HandshakeQueue {
+ public:
+  HandshakeQueue(size_t max_outstanding_handshakes)
+      : max_outstanding_handshakes_(max_outstanding_handshakes) {}
+
+  ~HandshakeQueue() { GRPC_COMBINER_UNREF(combiner_, "~HandshakeQueue"); }
+
+  void EnqueueHandshakeAndMaybeStartSome(alts_grpc_handshaker_client* client) {
+    Arg* arg = grpc_core::New<Arg>();
+    arg->client = client;
+    arg->self = this;
+    GRPC_CLOSURE_INIT(&arg->closure, EnqueueHandshakeAndMaybeStartSomeLocked,
+                      arg, nullptr);
+    combiner_->Run(&arg->closure, GRPC_ERROR_NONE);
+  }
+
+ private:
+  struct Node {
+    alts_grpc_handshaker_client* client = nullptr;
+    Node* next = nullptr;
+    Node* prev = nullptr;
+  };
+
+  struct Arg {
+    alts_grpc_handshaker_client* client;
+    HandshakeQueue* self;
+    grpc_closure closure;
+  };
+
+  static void EnqueueHandshakeAndMaybeStartSomeLocked(
+      void* arg, grpc_error* unused_error) {
+    Arg* args = static_cast<Arg*>(arg);
+    args->self->EnqueueHandshakeAndMaybeStartSomeInnerLocked(args->client);
+    grpc_core::Delete(args);
+  }
+
+  void EnqueueHandshakeAndMaybeStartSomeInnerLocked(
+      alts_grpc_handshaker_client* client) {
+    if (client == nullptr) {
+      outstanding_handshakes_--;
```

Comment:
Please add a comment explaining when `client` is `nullptr`.

Reply:
Done, I added more description on top of the `EnqueueHandshakeAndMaybeStartSome` API